fastestimator/pipeline.py

```diff
@@ -13,442 +13,132 @@
+            #num_process check
+            if self.num_process:
+                assert isinstance(self.num_process, int), "number of process must be integer or None"
+        elif isinstance(dataset, (DataLoader, tf.data.Dataset)):
+            if self.batch_size:
+                warnings.warn("batch_size will only be used for built-in dataset")
+            if self.ops:
+                warnings.warn("ops will only be used for built-in dataset")
+            if self.num_process:
+                warnings.warn("num_process will only be used for built-in dataset")
         else:
-            per_epoch.append(op)
-    return per_epoch
+            raise ValueError("Unsupported dataset type for {}".format(mode))
 
-
-class BasePipeline:
     def get_modes(self) -> Set[str]:
-        raise NotImplementedError()
-
-    def get_batch_size(self, mode: str, epoch: int) -> int:
-        raise NotImplementedError()
-
-    def get_signature_epochs(self, mode: str) -> List[int]:
-        raise NotImplementedError()
-
-    def get_all_output_keys(self) -> Set[str]:
-        raise NotImplementedError()
-
-    def get_num_examples(self, mode: str, epoch: int) -> int:
-        raise NotImplementedError()
-
-    def get_num_steps(self, mode: str, epoch: int) -> int:
-        return self.get_num_examples(mode=mode, epoch=epoch) // self.get_batch_size(mode=mode, epoch=0)
-
-    def get_iterator(self, mode: str, epoch: int = 0, dataset: Optional[Dataset] = None) -> Iterator:
-        raise NotImplementedError()
+        return set(mode for mode in self.data.keys())
 
     def benchmark(self, mode: str = "train", num_steps: int = 1000, log_interval: int = 100, epoch: int = 0):
-        itr = self.get_iterator(mode=mode, epoch=epoch)
+        loader = self.get_loader(mode=mode, epoch=epoch)
         start = time.perf_counter()
-        for idx in range(num_steps + 1):
-            _ = next(itr)
+        for idx, _ in enumerate(loader):
             if idx % log_interval == 0 and idx > 0:
                 duration = time.perf_counter() - start
-                batch_size = self.get_batch_size(mode=mode, epoch=epoch)
-                examples_per_sec = log_interval * batch_size / duration
-                print("FastEstimator: Step: {}, Epoch: {}, Batch Size: {}, Examples/sec: {}".format(
-                    idx, epoch, batch_size, examples_per_sec))
+                iters_per_sec = log_interval / duration
+                print("FastEstimator: Step: {}, Epoch: {}, Steps/sec: {}".format(idx, epoch, iters_per_sec))
                 start = time.perf_counter()
-
-    def show_results(self, mode: str = "train", num_steps: int = 1, epoch: int = 0):
-        data = []
-        itr = self.get_iterator(mode=mode, epoch=epoch)
-        for _ in range(num_steps):
-            data.append(next(itr))
+            if idx == num_steps:
+                break
+
+    def get_loader(self, mode: str, epoch: int = 0):
```

Comment:
typing, we should probably return an iterator rater than the loader since the loaders will be of different types

Reply:
no we can't because we cannot use for loop to access iterator multiple times.

Reply:
not sure what you mean, `for elem in iter` is pretty standard...

Reply:
try the following code and you will see:

```python
a = [1, 2, 3, 4]
it = iter(a)

#first time
for ele in it:
  print(ele)

#second time
for ele in it:
  print(ele)

```

Reply:
don't we invoke get_loader once per epoch though? why is that a use case for us?

Reply:
there are other parts of the code that performs some instance check on the loader.  we still can't return iterator.